### PR TITLE
EFSR-885 gender age plugins

### PR DESCRIPTION
**Pull request by z268**
Change list for the embedding calculator:
* removed facenet and insightface libraries from `embedding_calculator/srcext/`, now these install from pypi if necessary
* split insightface and facenet scanners into separate pluggins
* added PluginManager for finding plugins dependencies and importing plugins
* downloader for ML models from URL for plugins
* added age and gender recognition plugins from
  * insightface (mxnet)
  * ~rude_carnie (tensorflow)~
* API changes:
  * `/status` returns a list of availiable plugins with their short names
  * `/find_faces` accepts argument `face_plugins` with a comma-separated list of plugins short names (e.g. `face_plugins=calculator,age`
  * `/find_faces` returns execution time of each plugin on each detected face

### CODE REVIEW

File: embedding-calculator/src/docs/find_faces_post.yml [hunk 1]
```diff
@@ -2,7 +2,7 @@ tags:
   - Core
 summary: 'Find faces in the given image and return their bounding boxes.'
 description: 'Returns bounding boxes of detected faces on the image.'
-operationId: scanFacesPost
+operationId: findFacesPost
```
Comment (pospielov) [inline embedding-calculator/src/docs/find_faces_post.yml:5]: Is this swagger file? Then I don't see descriptions of gender and age plugins

File: embedding-calculator/src/services/facescan/plugins/facenet/facenet.py [hunk 2]
```diff
@@ -105,11 +99,43 @@ def find_faces(self, img: Array3D, det_prob_threshold: float = None) -> List[Bou
             filtered_bounding_boxes.append(box)
         return filtered_bounding_boxes
 
+
+class Calculator(base.BaseCalculator):
+    ml_models = (
+        # links from https://github.com/davidsandberg/facenet#pre-trained-models
+        # VGGFace2 training set, 0.9965 LFW accuracy
+        ('20180402-114759', 'https://drive.google.com/uc?id=1EXPBSXwTaqrSC0OhUdXNmKSh9qJUQ55-'),
```
Comment (pospielov) [inline embedding-calculator/src/services/facescan/plugins/facenet/facenet.py:107]: Are this models stored on our google drive? If not - let's put them to our google drive

File: embedding-calculator/src/services/facescan/plugins/insightface/insightface.py [hunk 3]
```diff
@@ -60,55 +53,43 @@ def scaled(self, coefficient: float) -> 'InsightFaceBoundingBox':
...
 
 
-class InsightFace(FaceScanner):
-    ID = 'InsightFace'
-    DETECTION_MODEL_NAME = ENV.DETECTION_MODEL
-    CALCULATION_MODEL_NAME = ENV.CALCULATION_MODEL
-    IMG_LENGTH_LIMIT = ENV.IMG_LENGTH_LIMIT
+class FaceDetector(InsightFaceMixin, base.BaseFaceDetector):
+    ml_models = (
+        ('retinaface_r50_v1', 'http://insightface.ai/files/models/retinaface_r50_v1.zip'),
```
Comment (pospielov) [inline embedding-calculator/src/services/facescan/plugins/insightface/insightface.py:77]: The same - let's store this models in our google drive
Or as an alternative - look at this solution - https://git-lfs.github.com/

File: embedding-calculator/src/services/facescan/plugins/insightface/insightface.py [hunk 4]
```diff
@@ -132,3 +113,56 @@ def find_faces(self, img: Array3D, det_prob_threshold: float = None) -> List[Ins
...
         return boxes
+
+    def crop_face(self, img: Array3D, box: InsightFaceBoundingBox) -> Array3D:
+        return face_align.norm_crop(img, landmark=box.landmark,
+                                    image_size=self.IMAGE_SIZE)
+
+
+class Calculator(InsightFaceMixin, base.BaseCalculator):
+    ml_models = (
+        ('arcface_r100_v1', 'http://insightface.ai/files/models/arcface_r100_v1.zip'),
```
Comment (pospielov) [inline embedding-calculator/src/services/facescan/plugins/insightface/insightface.py:124]: The same - let's store this models in our google drive

File: embedding-calculator/src/services/facescan/plugins/insightface/insightface.py [hunk 5]
```diff
@@ -132,3 +113,56 @@ def find_faces(self, img: Array3D, det_prob_threshold: float = None) -> List[Ins
...
+@attr.s(auto_attribs=True, frozen=True)
+class GenderAgeDTO(plugin_result.PluginResultDTO):
+    gender: str
+    age: Tuple[int, int]
+
+
+class GenderAgeDetector(InsightFaceMixin, base.BasePlugin):
+    slug = 'gender_age'
+    ml_models = (
+        ('genderage_v1', 'http://insightface.ai/files/models/genderage_v1.zip'),
```
Comment (pospielov) [inline embedding-calculator/src/services/facescan/plugins/insightface/insightface.py:153]: The same - let's store this models in our google drive

File: embedding-calculator/src/services/facescan/plugins/rude_carnie/rude_carnie.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+            best_i = int(np.argmax(output))
+            return labels[best_i], output[best_i]
+        return get_value
+
+
+class AgeDetector(base.BasePlugin):
+    slug = 'age'
+    LABELS = ((0, 2), (4, 6), (8, 12), (15, 20), (25, 32), (38, 43), (48, 53), (60, 100))
+    ml_models = (
+        ('22801', 'https://drive.google.com/uc?id=1JSggfO1FPu8eM1BeQ6yMG5nKrGbJDyqG'),
```
Comment (pospielov) [inline embedding-calculator/src/services/facescan/plugins/rude_carnie/rude_carnie.py:67]: The same - let's store this models in our google drive

File: embedding-calculator/src/services/facescan/plugins/rude_carnie/rude_carnie.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        model = _get_rude_carnie_model(self.LABELS, self.ml_model.path)
+        value, probability = model(face_img)
+        return plugin_result.AgeDTO(age=value, age_probability=probability)
+
+
+class GenderDetector(base.BasePlugin):
+    slug = 'gender'
+    LABELS = ('male', 'female')
+    ml_models = (
+        ('21936', 'https://drive.google.com/uc?id=1Gem2hM6bg746pqgTHQCyNv-Egf3CFgYf'),
```
Comment (pospielov) [inline embedding-calculator/src/services/facescan/plugins/rude_carnie/rude_carnie.py:80]: The same - let's store this models in our google drive

File: embedding-calculator/srcext/rude_carnie/README.md [hunk 8]
```diff
@@ -0,0 +1,227 @@
+Rude Carnie: Age and Gender Deep Learning with TensorFlow
```
Comment (pospielov) [inline embedding-calculator/srcext/rude_carnie/README.md:1]: We need to ask for permission or license before adding their code to our repository...

Comment (z268) [reply embedding-calculator/srcext/rude_carnie/README.md:1]: Asked - https://github.com/dpressel/rude-carnie/issues/109

File: embedding-calculator/src/docs/find_faces_post.yml (hunk 1 above)
Comment (z268) [reply embedding-calculator/src/docs/find_faces_post.yml:5]: It's just a fix of wrong name. Swagger description updated in PR #302

File: embedding-calculator/srcext/rude_carnie/README.md (hunk 8 above)
Comment (z268) [reply embedding-calculator/srcext/rude_carnie/README.md]: Removed this code until we get permissions

**Review by pospielov — APPROVED** [review]

**Pull request merged**